packages/sources/dlc-btc-por/src/transport/proof-of-reserves.ts

```diff
@@ -0,0 +1,219 @@
+import abi from '../config/dlc-manager-abi.json'
+import {
+  BitcoinNetwork,
+  BitcoinTransaction,
+  BitcoinTransactionVectorOutput,
+  getBitcoinNetwork,
+  RawVault,
+} from './utils'
+
+const logger = makeLogger('dlcBTC PoR')
+
+export type TransportTypes = BaseEndpointTypes
+
+export class DLCBTCPorTransport extends SubscriptionTransport<TransportTypes> {
+  name!: string
+  responseCache!: ResponseCache<TransportTypes>
+  requester!: Requester
+  provider!: ethers.providers.JsonRpcProvider
+  dlcManagerContract!: ethers.Contract
+  settings!: TransportTypes['Settings']
+
+  async initialize(
+    dependencies: TransportDependencies<TransportTypes>,
+    adapterSettings: TransportTypes['Settings'],
+    endpointName: string,
+    transportName: string,
+  ): Promise<void> {
+    await super.initialize(dependencies, adapterSettings, endpointName, transportName)
+    const { RPC_URL, CHAIN_ID, DLC_CONTRACT } = adapterSettings
+    this.settings = adapterSettings
+    this.provider = new ethers.providers.JsonRpcProvider(RPC_URL, CHAIN_ID)
+    this.dlcManagerContract = new ethers.Contract(DLC_CONTRACT, abi, this.provider)
+    this.requester = dependencies.requester
+  }
+  async backgroundHandler(context: EndpointContext<TransportTypes>) {
+    await this.handleRequest()
+    await sleep(context.adapterSettings.BACKGROUND_EXECUTE_MS)
+  }
+
+  async handleRequest() {
+    let response: AdapterResponse<TransportTypes['Response']>
+    try {
+      response = await this._handleRequest()
+    } catch (e) {
+      const errorMessage = e instanceof Error ? e.message : 'Unknown error occurred'
+      logger.error(e, errorMessage)
+      response = {
+        statusCode: 502,
+        errorMessage,
+        timestamps: {
+          providerDataRequestedUnixMs: 0,
+          providerDataReceivedUnixMs: 0,
+          providerIndicatedTimeUnixMs: undefined,
+        },
+      }
+    }
+    await this.responseCache.write(this.name, [{ params: [], response }])
+  }
+
+  async _handleRequest(): Promise<AdapterResponse<TransportTypes['Response']>> {
+    const providerDataRequestedUnixMs = Date.now()
+
+    // Get all vault data. Filter placeholder values.
+    const vaultData: RawVault[] = (await this.dlcManagerContract.getFundedDLCs(0, 10_000)).filter(
+      (v: RawVault) =>
+        v.uuid != '0x0000000000000000000000000000000000000000000000000000000000000000',
+    )
+
+    const depositsPromise = vaultData.map(async (vault) => {
+      try {
+        const isVerified = await this.verifyVaultDeposit(vault)
+        if (isVerified) {
+          return vault.valueLocked.toNumber()
+        }
+      } catch (e) {
+        logger.error(e, `Error while verifying Deposit for Vault: ${vault.uuid}. ${e}`)
+      }
+      return 0
+    })
+
+    // totalPoR represents total proof of reserves value in satoshis
+    const totalPoR = (await Promise.all(depositsPromise)).reduce((sum, deposit) => sum + deposit, 0)
```

Comment:
this is a really pretty solution, but assuming a case of 10000 DLCs, is it possible this would send 10000 parallel requests and overwhelm the RPC node?

Reply:
I think right now this is the only thing that concern me. (not necessarily right now, but maybe we face this issue in the next 6 month. again, depends on Joshua's answer). I lean towards adding a batch request support.

Reply:
Right, so there are currently 2 vaults in mainnet, right?
Can we just multiply each by 100x or 1000x here and send to one of our RPCs and see if it falls over? 😂

Reply:
I'll try that, but i will implement batching anyway